GVFS/GVFS.Service/Mac/GVFSService.cs

```diff
@@ -0,0 +1,115 @@
+using GVFS.Common;
+using GVFS.Common.NamedPipes;
+using GVFS.Common.Tracing;
+using System;
+using System.Linq;
+using System.Threading;
+
+namespace GVFS.Service
+{
+    public class GVFSService
+    {
+        private const string ServiceNameArgPrefix = "--servicename=";
+        private const string EtwArea = nameof(GVFSService);
+
+        private JsonTracer tracer;
+        private Thread serviceThread;
+        private ManualResetEvent serviceStopped;
+        private string serviceName;
+        private string serviceDataLocation;
+
+        public GVFSService(JsonTracer tracer)
+        {
+            this.tracer = tracer;
+            this.serviceName = GVFSConstants.Service.ServiceName;
+        }
+
+        public void RunWithArgs(string[] args)
+        {
+            if (this.serviceThread != null)
+            {
+                throw new InvalidOperationException("Cannot start service twice in a row.");
+            }
+
+            string nameArg = args.FirstOrDefault(arg => arg.StartsWith(ServiceNameArgPrefix, StringComparison.OrdinalIgnoreCase));
+            if (nameArg != null)
+            {
+                this.serviceName = nameArg.Substring(ServiceNameArgPrefix.Length);
+            }
+
+            try
+            {
+                this.serviceDataLocation = Paths.GetServiceDataRoot(this.serviceName);
+                this.StartServiceThread();
+                this.serviceThread.Join();
+            }
+            catch (Exception e)
+            {
+                this.LogExceptionAndExit(e, nameof(this.RunWithArgs));
+            }
+        }
+
+        private void StartServiceThread()
+        {
+            if (this.serviceStopped != null)
+            {
+                return;
+            }
+
+            this.serviceStopped = new ManualResetEvent(false);

```

Comment:
Rather than having this code here:

```
if (this.serviceStopped != null)
{
    return;
}

this.serviceStopped = new ManualResetEvent(false);
```

Can we just construct the `ManualResetEvent` in the constructor?

(I am not quite sure why the `if (this.serviceStopped != null)` check is there right now)

Reply:
Moved the creation to constructor.

On Windows I think this check `if (this.serviceStopped != null)` helps avoid the service thread from being started multiple times. This pattern does not apply on the Mac - since launchd only starts the service once (there is no suspend/resume)